### PR TITLE
feat: add command validation for user types and game phases

### PR DESCRIPTION
**Pull request by ratteler50**
This pull request introduces enhancements to the command registry system, focusing on validation mechanisms and testing. The key changes include adding user type and game phase validation for commands, updating type annotations for consistency, and expanding test coverage to ensure the new validation logic works as expected.

### Enhancements to Command Validation:
* Added a `ValidationError` exception class for handling validation failures. (`commands/registry.py`, [commands/registry.pyL4-R22](diffhunk://#diff-3986cf384adebd590bffd8106f5b1fa609a1b04401e55bc78a8b3707818e2a7eL4-R22))
* Implemented `validate_user_type` and `validate_game_phase` functions to enforce permissions based on user roles and game phases. (`commands/registry.py`, [commands/registry.pyR114-R220](diffhunk://#diff-3986cf384adebd590bffd8106f5b1fa609a1b04401e55bc78a8b3707818e2a7eR114-R220))
* Integrated validation checks into the `handle_command` method to ensure commands are executed only if the user meets the required criteria. (`commands/registry.py`, [commands/registry.pyL205-R345](diffhunk://#diff-3986cf384adebd590bffd8106f5b1fa609a1b04401e55bc78a8b3707818e2a7eL205-R345))

### Type Annotations and Code Consistency:
* Updated type annotations for lists to use Python's modern `list[...]` syntax instead of `List[...]`. (`commands/registry.py`, [commands/registry.pyL110-R235](diffhunk://#diff-3986cf384adebd590bffd8106f5b1fa609a1b04401e55bc78a8b3707818e2a7eL110-R235))

### Test Suite Expansion:
* Added a new test module, `test_enum_enforcement.py`, to validate user type and game phase restrictions, including edge cases for public users and inactive games. (`tests/commands/test_enum_enforcement.py`, [tests/commands/test_enum_enforcement.pyR1-R243](diffhunk://#diff-3ac2d7f502824b0d3d80fe843b6e7185a2ab4d86d3644fa14df223f5f8ff701cR1-R243))
* Updated existing tests to use a `storyteller` user instead of `alice` for better alignment with the new validation logic. (`tests/commands/test_debug_commands_registry.py`, [[1]](diffhunk://#diff-98a67aa0797d2a746cc4c151206658be86eb861d915faeda34639cec4dd6f62dL56-R57) [[2]](diffhunk://#diff-98a67aa0797d2a746cc4c151206658be86eb861d915faeda34639cec4dd6f62dL83-R84) [[3]](diffhunk://#diff-98a67aa0797d2a746cc4c151206658be86eb861d915faeda34639cec4dd6f62dL180-R181)

### CODE REVIEW

File: commands/registry.py [hunk 1]
```diff
@@ -99,6 +111,113 @@ def _format_arg(arg: CommandArgument) -> str:
...
+    # Check if user matches any of the required types
+    user_matches = False
+
+    if UserType.STORYTELLER in command_info.user_types and is_storyteller:
+        user_matches = True
+    elif UserType.PLAYER in command_info.user_types and is_player:
+        user_matches = True
+    elif UserType.OBSERVER in command_info.user_types and is_observer:
+        user_matches = True
+    elif UserType.PUBLIC in command_info.user_types and not (is_storyteller or is_player or is_observer):
```
Comment (jking-roar) [inline commands/registry.py:168]: What's an example of a command that's public?  Is this restricted so that only those people who are not st/player/observer can use it?

File: commands/registry.py [hunk 2]
```diff
@@ -99,6 +111,113 @@ def _format_arg(arg: CommandArgument) -> str:
...
+    role_name_map = {
+        UserType.STORYTELLER: "Storyteller",
+        UserType.PLAYER: "Player",
+        UserType.OBSERVER: "Observer",
+        UserType.PUBLIC: "Public",
+    }
+    role_names = [role_name_map.get(user_type, str(user_type)) for user_type in command_info.user_types]
+    required_roles = ", ".join(role_names)
+    raise ValidationError(
+        f"You do not have permission to use the {command_info.name} command. Required role(s): {required_roles}.")
```
Comment (jking-roar) [inline commands/registry.py:184]: One command was added in the past to support enhanced voting power but intentionally hidden from the response output if a player tried using it. The goal is to prevent players from discovering it by trying the command and receiving a specific error that reveals its existence. (even though the code was public :see_no_evil:)

Is there a way to make the fallback behavior more generic—so that unimplemented or hidden commands don’t leak information about their presence?

File: commands/registry.py [hunk 3]
```diff
@@ -99,6 +111,113 @@ def _format_arg(arg: CommandArgument) -> str:
...
+
```
Comment (jking-roar) [inline commands/registry.py:185]: should be "Allowed" roles rather than required.  if there's multiple, required suggests that you need both roles

File: commands/registry.py [hunk 4]
```diff
@@ -107,13 +226,13 @@ def __init__(self):
         self.aliases: Dict[str, str] = {}
 
     def command(self, name: str,
-                aliases: Optional[List[str]] = None,
-                user_types: Optional[List[UserType]] = None,
-                arguments: Union[List[CommandArgument], Dict[UserType, List[CommandArgument]]] = None,
+                aliases: Optional[list[str]] = None,
```
Comment (jking-roar) [inline commands/registry.py:229]: Is there a behavior change in this typing change from List to list?

File: commands/registry.py (hunk 1 above)
Comment (ratteler50) [reply commands/registry.py:168]: So this was a design choice I was unsure of how to handle.

The way that I see it is you partition everyone on the server in four:
Soryteller, Observer, Player, and everyone else.

There are commands that should be restricted to people actively involved in the game but no specific type (e.g. history), and commands that should not be restricted at all (e.g. `help`).

There is no use case I can think of that is _only_ available to public, but there are use cases where we want to allow anyone on the server vs restricting from people not actively involved in the game)

I don't expect there to be any commands that are _only_ public, but _adding_ public to the list means public will be allowed to call it and not adding it means public will not.

Another option is to have a fully empty list be the same as fully public (currently it means nobody can run the command which also we probably won't do in practice), and then have the existance of the list mean "restricted to" and absense of the list mean "no restrictions".

The main reason I didn't go that way is it felt too easy to accidentaly grant wider permissions than intended which isn't great, "failing closed" felt safer her.  It also might complicate the implementation a bit but that's a one time kinda local thing. Talking it through now it might not be a big deal and maybe we just want empty list to mean the same thing as [STORYTELLER, OBSERVER, PLAYER, PUBLIC] currently does.

Thoughts?

File: commands/registry.py (hunk 2 above)
Comment (ratteler50) [reply commands/registry.py:184]: Hmm I'll give that some thought.  You can definitely hide the command from the HELP list by not putting it in any help section, then it won't show up when someone runs help.

It's hacky, but the easiest solution to the case you're suggesting would be to mark the command as being allowed to run by anyone in any gamestate, and doing manual check for being ST in the implementation itself and a generic "command not found" repsonse if not.

Basically exactly what you did with the current implementation:
```
if (
    global_vars.game is NULL_GAME
    or global_vars.gamemaster_role
    not in global_vars.server.get_member(message.author.id).roles
):
    await message_utils.safe_send(
        message.author,
        "Command {} not recognized. For a list of commands, type @help.".format(
            command
        ),
    )

```

File: commands/registry.py (hunk 4 above)
Comment (ratteler50) [reply commands/registry.py:229]: Not for python 3.9+ (and we're python 3.10).

It's a modernization suggested for 3.9+ from using the Type module to using the builtin (which became supported in 3.9), and we are already using `list[type_in_list]` elsewhere which is the thing added in 3.9.


I did just learn while looking up which version of python the change was that even though typing hints are only for static analysis and not used at runtime (e.g. I could change `Optional[list[str]]` to `tuple[int]` and my IDE would throw a fit but it wouldn't affect runtime at all unless you turned on additional tools to enforce it), pre 3.9 would fail to parse this so wouldn't run at all even though it's only used for static analysis so would be an issue (but we're not pre 3.9 so it's not an issue here)

File: commands/registry.py (hunk 1 above)
Comment (ratteler50) [reply commands/registry.py:168]: fwiw I did an analysis of the current permission distribution of current commands and only like 3 are accessible to public so seemed like it makes more sense to be the more verbose option than the less verbose option

**Review by jking-roar — APPROVED** [review]
